File: Popularity/DCAFPilot/src/python/DCAF/services/sitedb.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-
#pylint: disable=
"""
File       : sitedb.py
Author     : Valentin Kuznetsov <vkuznet AT gmail dot com>
Description: SiteDB service module
"""

# system modules
import time
from   types import InstanceType

# package modules
import DCAF.utils.jsonwrapper as json
from DCAF.utils.utils import genkey
from DCAF.services.generic import GenericService

def rowdict(columns, row):
    """Convert given row list into dict with column keys"""
    robj = {}
    for key, val in zip(columns, row):
        robj.setdefault(key, val)
    return robj
    
def sitedb_parser(source):
    """SiteDB parser"""
    if  isinstance(source, str) or isinstance(source, unicode):
        data = json.loads(source)
    elif isinstance(source, InstanceType) or isinstance(source, file):
        # got data descriptor
        try:
            data = json.load(source)
        except Exception as exc:
            print_exc(exc)
            source.close()
            raise
        source.close()
    else:
        data = source
    if  not isinstance(data, dict):
        raise Exception('Wrong data type, %s' % type(data))
    if  'desc' in data:
        columns = data['desc']['columns']
        for row in data['result']:
            yield rowdict(columns, row)
    else:
        for row in data['result']:
            yield row

def user_data(username, sitedb_dict):
    "Find user data in SiteDB dict"
    columns = sitedb_dict['desc']['columns']
    for row in sitedb_dict['result']:
        idx = columns.index('username')
        if  row[idx] == username:
            return dict(zip(columns, row))

class SiteDBService(GenericService):
    """
    Helper class to provide DBS service
    """
    def __init__(self, config=None, verbose=0):
        GenericService.__init__(self, config, verbose)
        self.name = 'sitedb'
        self.url = config['services'][self.name]

    def fetch(self, api, params=None):
        "Fetch data for given api"
        if  api == 'sites':
            api = 'site-names'
        url = '%s/%s' % (self.url, api)
        data = super(SiteDBService, self).fetch(url, params)
        for row in sitedb_parser(data):
            if  api == 'people':
                rid = genkey(str(row['dn']), truncate=5)
                rec = {'dn':row['dn'], 'rid':rid}
            if  api == 'site-names':
                rid = genkey(str(row['alias']), truncate=5)
                rec = {'site':row['alias'], 'rid':rid}
            yield rec

    def update(self, cname):
        "Update internal database with fresh snapshot of data"
        if  self.verbose:
            print "%s update %s" % (self.name, cname)
        self.storage.cleanup(cname)
        docs = self.fetch(cname)
        self.storage.insert(cname, docs)

    def dnid(self, dn):
        "Return DN id from dn database"
        spec = {'dn':dn}
        res = [r for r in self.storage.fetch('people', spec)]
        if  not res:
            rid = self.storage.count('people', {})+1
            self.storage.insert('people', [{'dn':dn,'rid':rid}])
            return rid
        doc = [r for r in res][0]
        return doc['rid']

    def siteid(self, site):
        "Return site id from site database"
        spec = {'site':site}
        res = [r for r in self.storage.fetch('sites', spec)]
        if  not res:
            rid = self.storage.count('sites', {})+1
            self.storage.insert('sites', [{'site':dn,'rid':rid}])
            return rid
        doc = [r for r in res][0]
        return doc['rid']

File: Popularity/DCAFPilot/src/Go/dataframe/src/cms/sitedb.go
/*
 *
 * Author     : Valentin Kuznetsov <vkuznet AT gmail dot com>
 * Description: part of cms package responsible for sitedb calls
 * Created    : Wed Feb 10 19:31:44 EST 2016
 */
package cms

import (
	"encoding/json"
	"fmt"
	"strings"
	"utils"
)

// helper function to load SiteDB data stream
func loadSiteDBData(furl string, data []byte) []Record {
	var out []Record
	var rec Record
	err := json.Unmarshal(data, &rec)
	if err != nil {
		if utils.VERBOSE > 0 {
			msg := fmt.Sprintf("SiteDB unable to unmarshal the data, furl=%s, data=%s, error=%v", furl, string(data), err)
			fmt.Println(msg)
		}
		return out
	}
	desc := rec["desc"].(map[string]interface{})
	headers := desc["columns"].([]interface{})
	values := rec["result"].([]interface{})
	for _, item := range values {
		row := make(Record)
		val := item.([]interface{})
		for i, h := range headers {
			key := h.(string)
			row[key] = val[i]
			if key == "username" {
				row["name"] = row[key]
			}
		}
		out = append(out, row)
	}
	return out
}

// helper function to get CMS site names for given siteName
func siteNames(site string) []string {
	var out []string
	api := "site-names"
	furl := fmt.Sprintf("%s/%s", sitedbUrl(), api)
	if utils.VERBOSE > 1 {
		fmt.Println("furl", furl)
	}
	response := utils.FetchResponse(furl, "")
	if response.Error == nil {
		records := loadSiteDBData(furl, response.Data)
		for _, r := range records {
			siteName := r["alias"].(string)
			siteType := r["type"].(string)
			if siteType == "phedex" && strings.HasPrefix(siteName, site) {
				if strings.HasPrefix(site, "T1_") {
					if strings.HasSuffix(siteName, "_Disk") {
						out = append(out, siteName)
					}
				} else {
					out = append(out, siteName)
				}
			}
		}
	}
	return utils.List2Set(out)
}
